packages/medusa/src/repositories/product.ts

```diff
@@ -1,21 +1,408 @@
+  "relations"
+>
+
+export type FindWithoutRelationsOptions = DefaultWithoutRelations & {
+  where: DefaultWithoutRelations["where"] & {
+    price_list_id?: FindOperator<PriceList>
+    sales_channel_id?: FindOperator<SalesChannel>
+    category_id?: {
+      value: string[]
+    }
+    categories?: FindOptionsWhere<ProductCategory>
+    tags?: FindOperator<ProductTag>
+    include_category_children?: boolean
+    discount_condition_id?: string
+  }
+}
+
 export const ProductRepository = dataSource.getRepository(Product).extend({
+  mergeEntitiesWithRelations(
+    entitiesAndRelations: Array<Partial<Product>>
+  ): Product[] {
+    const entitiesAndRelationsById = groupBy(entitiesAndRelations, "id")
+    return map(entitiesAndRelationsById, (entityAndRelations) =>
+      merge({}, ...entityAndRelations)
+    )
+  },
+
+  async queryProducts(
+    optionsWithoutRelations: FindWithoutRelationsOptions,
+    shouldCount = false
+  ): Promise<[Product[], number]> {
+    const productAlias = "product"
+
+    const tags = optionsWithoutRelations?.where?.tags
+    delete optionsWithoutRelations?.where?.tags
+
+    const price_lists = optionsWithoutRelations?.where?.price_list_id
+    delete optionsWithoutRelations?.where?.price_list_id
+
+    const sales_channels = optionsWithoutRelations?.where?.sales_channel_id
+    delete optionsWithoutRelations?.where?.sales_channel_id
+
+    const categoryId = optionsWithoutRelations?.where?.category_id
+    delete optionsWithoutRelations?.where?.category_id
+
+    const categories = optionsWithoutRelations?.where?.categories
+    delete optionsWithoutRelations?.where?.categories
+
+    const include_category_children =
+      optionsWithoutRelations?.where?.include_category_children
+    delete optionsWithoutRelations?.where?.include_category_children
+
+    const discount_condition_id =
+      optionsWithoutRelations?.where?.discount_condition_id
+    delete optionsWithoutRelations?.where?.discount_condition_id
+
+    const qb = this.createQueryBuilder(productAlias)
+      .select([`${productAlias}.id`])
+      .skip(optionsWithoutRelations.skip)
+      .take(optionsWithoutRelations.take)
+
+    if (optionsWithoutRelations.where) {
+      qb.where(optionsWithoutRelations.where)
+    }
+
+    if (tags) {
+      qb.leftJoin(`${productAlias}.tags`, "tags").andWhere(
+        `tags.id IN (:...tag_ids)`,
+        {
+          tag_ids: tags.value,
+        }
+      )
+    }
+
+    if (price_lists) {
+      qb.leftJoin(`${productAlias}.variants`, "variants")
+        .leftJoin("variants.prices", "prices")
+        .andWhere("prices.price_list_id IN (:...price_list_ids)", {
+          price_list_ids: price_lists.value,
+        })
+    }
+
+    if (sales_channels) {
+      qb.innerJoin(
+        `${productAlias}.sales_channels`,
+        "sales_channels",
+        "sales_channels.id IN (:...sales_channels_ids)",
+        { sales_channels_ids: sales_channels.value }
+      )
+    }
+
+    /* if (false && categories) {
```

Comment:
@riqwan this is the part we will have to look together, when filtering the categories